traits/tests/test_regression.py

```diff
@@ -289,3 +300,140 @@ class A(HasTraits):
+        except Exception:
+            self.fail("Mutating a nested list should not fail.")
+
+    def test_modify_list_in_list_no_events(self):
+        instance = NestedContainerClass(list_of_list=[[]])
+        instance.on_trait_change(self.change_handler, "list_of_list_items")
+
+        instance.list_of_list[0].append(1)
+        self.assertEqual(len(self.events), 0, "Expected no events.")
+
+    def test_modify_dict_in_list(self):
+        instance = NestedContainerClass(list_of_dict=[{}])
+
+        try:
+            instance.list_of_dict[0]["key"] = 1
+        except Exception:
+            self.fail("Mutating a nested dict should not fail.")
+
+    def test_modify_dict_in_list_with_new_value(self):
+        instance = NestedContainerClass(list_of_dict=[{}])
+
+        instance.list_of_dict.append(dict())
+        try:
+            instance.list_of_dict[-1]["key"] = 1
+        except Exception:
+            self.fail("Mutating a nested dict should not fail.")
+
+    def test_modify_dict_in_dict_does_not_fire(self):
+        # Related to enthought/traits#25
+        instance = NestedContainerClass(dict_of_dict={"1": {"2": 2}})
+        instance.on_trait_change(self.change_handler, "dict_of_dict_items")
+
+        instance.dict_of_dict["1"]["3"] = 3
+
+        self.assertEqual(len(self.events), 0)
+
+    def test_modify_dict_in_union_in_dict(self):
+        instance = NestedContainerClass(
+            dict_of_union_none_or_dict={"1": {"2": 2}},
+        )
+        instance.on_trait_change(
+            self.change_handler, "dict_of_union_none_or_dict_items")
+
+        instance.dict_of_union_none_or_dict["1"]["3"] = 3
+
+        self.assertEqual(len(self.events), 0)
+
+    def test_modify_set_in_list(self):
+        instance = NestedContainerClass(list_of_set=[set()])
+        try:
+            instance.list_of_set[0].add(1)
+        except Exception:
+            self.fail("Mutating a nested set should not fail.")
+
+    def test_modify_set_in_list_with_new_value(self):
+        instance = NestedContainerClass(list_of_set=[])
+        instance.list_of_set.append(set())
+        try:
+            instance.list_of_set[0].add(1)
+        except Exception:
+            self.fail("Mutating a nested set should not fail.")
+
+    def test_modify_set_in_dict_no_events(self):
+        instance = NestedContainerClass(dict_of_set={"1": set()})
+        instance.on_trait_change(self.change_handler, "dict_of_set_items")
+
+        instance.dict_of_set["1"].add(1)
+
+        self.assertEqual(len(self.events), 0, "Expected no events.")
+
+    def test_modify_set_in_union_in_dict(self):
+        instance = NestedContainerClass(
+            dict_of_union_none_or_set={"1": set()}
+        )
+        instance.on_trait_change(
+            self.change_handler, "dict_of_union_none_or_set_items")
+
+        instance.dict_of_union_none_or_set["1"].add(1)
+
+        self.assertEqual(len(self.events), 0, "Expected no events.")
```

Comment:
I think these two tests should check that the modification happens without error rather than that no events have been fired. The way it is right now, a regression would result in `ERROR` rather than `FAIL` (as seen in test failures at c91e91e)

Reply:
Good point: the test objectives are a bit confusing here. I will make the above two tests to check mutation does not fail and I will add a separate test for no events, as this issue could have been fixed differently such that the mutation goes through and an event is fired. 
Unlike list and dict where I could have list of list and dict of dict to just test the nonexistence of event without running into an error at the time of mutation, I can't have set of set because set is not hashable. Therefore I will need to borrow the dict or list to be that outer container.  Since the assertion for no events is the test objective, I am afraid the regression will have to cause a test error for that test.